src/merkle/store/tests.rs

```diff
@@ -647,6 +647,70 @@ fn test_constructors() -> Result<(), MerkleError> {
     Ok(())
 }
 
+#[test]
+fn node_path_should_be_truncated_by_midtier_insert() {
+    let key = 0b11010010_11001100_11001100_11001100_11001100_11001100_11001100_11001100_u64;
+
+    let mut store = MerkleStore::new();
+    let root: Word = EmptySubtreeRoots::empty_hashes(64)[0].into();
+
+    // insert first node - works as expected
+    let depth = 64;
+    let node = [Felt::new(key); WORD_SIZE];
+    let index = NodeIndex::new(depth, key).unwrap();
+    let root = store.set_node(root, index, node).unwrap().root;
+    let result = store.get_node(root, index).unwrap();
+    let path = store.get_path(root, index).unwrap().path;
+    assert_eq!(node, result);
+    assert_eq!(path.depth(), depth);
+    assert!(path.verify(index.value(), result, &root));
+
+    // flip the first bit of the key and insert the second node on a different depth
+    let key = key ^ (1 << 63);
+    let key = key >> 8;
+    let depth = 56;
+    let node = [Felt::new(key); WORD_SIZE];
+    let index = NodeIndex::new(depth, key).unwrap();
+    let root = store.set_node(root, index, node).unwrap().root;
+    let result = store.get_node(root, index).unwrap();
+    let path = store.get_path(root, index).unwrap().path;
+    assert_eq!(node, result);
+    assert_eq!(path.depth(), depth);
+    assert!(path.verify(index.value(), result, &root));
+
+    // attempt to fetch a path of the second node to depth 64
+    // should fail because the previously inserted node will remove its sub-tree from the set
+    let key = key << 8;
+    let index = NodeIndex::new(64, key).unwrap();
+    assert!(store.get_node(root, index).is_err());
+}
+
+#[test]
+fn get_leaf_depth_works() {
+    let key = u64::MAX;
```

Comment:
nit: maybe use a different key with mixed true/false bits?

Reply:
The intent of that test was to produce the following effect

![image](https://user-images.githubusercontent.com/8730839/230188642-96ca955a-bf08-45c6-abe6-c54cce3ef8ca.png)

But it was written in a very confusing way, so I rewrote it. Basically we have a full right path, and then we flip the  first bit for every iteration. I'll also add a variant that will flip and shift down so we test all depths

Reply:
I'll add a couple of tests now with declared cases so we demonstrate the routine is working

Reply:
I added two tests: one that will perform the rainbow check described above, the other who will try a couple of hardcoded leaves and expect the correct behavior.